easypy/threadtree.py

```diff
@@ -170,27 +177,96 @@ def iter_thread_frames():
+    to the wrapped frame object, making the wrapper behave like the original frame.
+    """
+
+    def __init__(self, frame, thread_ident):
+        self._frame = frame
+        self.f_thread_ident = thread_ident
+
+    def __getattr__(self, name):
+        """Delegate attribute access to the wrapped frame object."""
+        return getattr(self._frame, name)
+
+    def __setattr__(self, name, value):
+        """Allow setting attributes."""
+        if name == "_frame" or name == "f_thread_ident":
+            super().__setattr__(name, value)  # Handle internal attributes
+        else:
+            setattr(self._frame, name, value)  # Delegate to the wrapped frame
+
+    def __repr__(self):
+        """Provide a string representation including the class name and f_thread_ident."""
+        return (
+            f"<{self.__class__.__name__}(frame={repr(self._frame)}, "
+            f"f_thread_ident={self.f_thread_ident})>"
+        )
+
+    def __str__(self):
+        """Provide a readable string for the wrapper."""
+        return str(self._frame)
+
+    def __eq__(self, other):
+        """Equality comparison."""
+        if isinstance(other, self.__class__):
+            return (
+                self._frame == other._frame
+                and self.f_thread_ident == other.f_thread_ident
+            )
+        return self._frame == other
+
+    def __ne__(self, other):
+        """Inequality comparison."""
+        return not self.__eq__(other)
+
+    def __hash__(self):
+        """Make the object hashable."""
+        return hash(self._frame)
+
+    def __dir__(self):
+        """List attributes of the wrapper and wrapped object."""
+        return dir(self._frame) + ["f_thread_ident"]
+
+
+def walk_frames(thread=None, *, across_threads=False, use_snapshots=False):
     """
     Yields the stack frames of the current/specified thread.
 
     :param across_threads: yield frames of ancestor threads.
-    Note that the parent thread might be off to other things, and not actually in the frame that spawned the thread at the tip.
+    :param use_snapshots: If True, yield frames from thread snapshots, ensuring consistent frame retrieval even if the
+    parent thread has moved on.
+
+    This function traverses the stack of the given thread, optionally walking up the stack of ancestor threads if
+    `across_threads` is set.
+    When `use_snapshots` is enabled, frames are retrieved from snapshots, mitigating issues where the parent thread
+    might no longer be at the frame that spawned the current thread.
     """
 
     if not thread:
         thread = threading.current_thread()
 
-    frame = dict(iter_thread_frames()).get(thread.ident)
-
-    while frame:
-
-        yield frame
-
-        frame = frame.f_back
-        if not frame and across_threads and thread.parent:
-            thread = thread.parent
-            frame = dict(iter_thread_frames()).get(thread.ident)
+    if use_snapshots:
```

Comment:
Traversing across threads with `use_snapshots=True` is around 65 times faster than without it
<img width="1696" alt="Use snapshots" src="https://github.com/user-attachments/assets/9c2489aa-6636-4793-8c8f-560736c7617f" />

Reply:
That's further argument to drop the `use_snapshots` flag and make it the default behavior